### PR TITLE
Feature/pdf download instructions

### LINKED ISSUE #247: add pdf download for instructions
create screenshot / renderer of each layer, combine into pdf, offer download

### PR DESCRIPTION
**Pull request by AtiX**
first version of something. not pdf, but html is nice, too. closes #247 
fixes several smaller/larger bugs in download functionality,
also improves responsiveness of Export UI.

### CODE REVIEW

**Comment by stuikomma** [comment]
highlightVoxel is visible in renderings

**Comment by stuikomma** [comment]
provide download on landingpage

**Comment by stuikomma** [comment]
Disable download initialization in left bundle

**Comment by AtiX** [comment]
Fixed instruction image size - it's now independent from current viewport and added openSCAD export and part list. Please review!

File: src/client/landingpage.coffee [hunk 1]
```diff
@@ -33,12 +33,22 @@ globalConfig.rendering.usePipeline = false
...
 config1.plugins.newBrickator = false
+config1.plugins.legoInstructions = false
 
 # configure right bundle to not show the model
 config2.showModel = false
 
+# configure right bundle to offer downloading lego instructions
+config2.offerDownload = true
+config2.downloadSettings = {
+	testStrip: false
```
Comment (Steditor) [inline src/client/landingpage.coffee]: Is it a good idea to deactivate the teststrip on the landingpage? I would show it....

File: src/plugins/legoInstructions/OpenScadGenerator.coffee [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		 * \n
+		 * Brick layout for openSCAD\n
+		 * Generated with http://brickify.it\n
+		 *\n
+		 */\n\n
+	'
+
+_scadBase = ->
+	return '
+		gridSpacing = [8, 8, 3.2];
```
Comment (Steditor) [inline src/plugins/legoInstructions/OpenScadGenerator.coffee]: Can we maybe extract this code to a standard .scad file and require it here?

File: src/plugins/legoInstructions/PartListGenerator.coffee [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	# is always first
+	for part in partList
+		if part.size.x > part.size.y
+			tmp = part.size.x
+			part.size.x = part.size.y
+			part.size.y = tmp
+
+	return partList
+
+equalSize = (a,b) ->
```
Comment (Steditor) [inline src/plugins/legoInstructions/PartListGenerator.coffee]: I would put this as static method in `Brick.coffee`.

File: src/plugins/legoInstructions/PartListGenerator.coffee [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	return ((a.x == b.x and a.y == b.y) or
+		(a.x == b.y and a.y == b.x)) and a.z == b.z
+
+module.exports.getHtml = (list) ->
+	html = '<h3>Bricks needed</h3>'
+	html += '<p>This is a list of how many and what types of'
+	html += ' bricks you need to build this model: <ul>'
+
+	for part in list
+		html += '<li>'
```
Comment (Steditor) [inline src/plugins/legoInstructions/PartListGenerator.coffee]: I would suggest using a table instead of a list.

File: src/common/globals.yaml [hunk 5]
```diff
@@ -92,3 +94,12 @@ gridSpacing:
...
 exportStepSize: 0.033
+
+downloadSettings:
+  testStrip: true
+  stl: true
+  lego: true
+  steps: 4
+
+# resolution of the images in the build instructions
+legoInstructionResolution: 1024
```
Comment (Steditor) [inline src/common/globals.yaml]: Missing new line at EOF

**Comment by stuikomma** [comment]
:boat:, pls recheck (e.g. @Steditor ?) :wink:

**Comment by Steditor** [comment]
My download file is named "brickify-googles.stl-pdf.zip" and contains .html, .png and .scad files -
 - [ ] clean up file name extensions

**Comment by Steditor** [comment]
On printing the instructions are very tattered - we should use [page-break-props](http://www.w3.org/TR/CSS21/page.html#page-break-props)

**Comment by Steditor** [comment]
Maybe it would be good to separate the steps better. It's not always clear whether "Layer X" is a headline or a subtitle of the image.

**Comment by Steditor** [comment]
The rendering quality changes within the instructions.
 - [ ] render with one fixed quality (e.g. DH or PH)

**Comment by Steditor** [comment]
See #574

**Comment by Steditor** [comment]
Blocked by #582

**Comment by Steditor** [comment]
Progress (#582) will be done with the new download dialog (#608)

**Comment by Steditor** [comment]
Closes #558

**Comment by AtiX** [comment]
:rocket:

**Pull request merged**